src/coreclr/jit/emit.cpp

```diff
@@ -1005,19 +1005,20 @@ insGroup* emitter::emitSavIG(bool emitAdd)
 
     if (sz != 0)
     {
-        assert(emitLastIns != nullptr);
-        assert(emitCurIGfreeBase <= (BYTE*)emitLastIns);
-        assert((BYTE*)emitLastIns < emitCurIGfreeBase + sz);
+        assert(emitHasLastIns());
+        assert(emitCurIGfreeBase <= (BYTE*)emitGetLastIns());
+        assert((BYTE*)emitGetLastIns() < emitCurIGfreeBase + sz);
 
 #if defined(TARGET_XARCH)
-        assert(emitLastIns != nullptr);
-        if (emitLastIns->idIns() == INS_jmp)
+        assert(emitHasLastIns());
+        if (emitGetLastIns()->idIns() == INS_jmp)
         {
             ig->igFlags |= IGF_HAS_REMOVABLE_JMP;
         }
 #endif
 
-        emitLastIns = (instrDesc*)((BYTE*)id + ((BYTE*)emitLastIns - (BYTE*)emitCurIGfreeBase));
+        emitLastInstrs[emitInsCount % ArrLen(emitLastInstrs)] =
+            (instrDesc*)((BYTE*)id + ((BYTE*)emitGetLastIns() - (BYTE*)emitCurIGfreeBase));
```

Comment:
There's a bug here:

We fill a static buffer with instrDesc's. When the buffer is full, or we move to a new BasicBlock with a label, we save that by allocating memory and moving them to new memory. This code updates the `emitLastIns` pointer to point to that new memory. Your new code only moves the last one, but doesn't update all the other instructions that have been moved. You need some kind of loop of size `emitCurIGinsCnt` to move them all.

Reply:
This might make this PR more complicated. If I need to update all the previous instructions, I have a feeling that TP is going to suffer. We should discuss this offline.

Reply:
Hhm, my impression was that whenever we see new IG, we would reset the `lastIns` table (to make sure TP is not impacted because of this, it was planned to just keep reset the current pointer and the no. of valid entries in the table and overwrite existing table), isn't that the case? I had related question around that - if my understanding is correct, how reliably can we check about `emitHasLastIns()` , etc.?

Reply:
Also, you won't be checking cross IG last instructions, so why should we update this to point to new buffer?

Reply:
I updated the PR to update the last instructions to point to the new memory, the TP diff is bad:

After: with updating
> [11:14:22] Total instructions executed delta: 144653690 (0.12% of base)

vs.

Before: without updating
> [11:17:27] Total instructions executed delta: 22510491 (0.02% of base)

--

I'm going to look into other ways to handle this. We really can't afford to update all instructions - instead, when we save the IG, we should just reset what is allowed to be enumerated in the last instrs array.

Reply:
Trying something different, I added `emitCurLastInsCnt` to keep track of the count. When we update the last instr here, I reset the count to `1`.